dpctl/tensor/libtensor/include/kernels/elementwise_functions/round.hpp

```diff
@@ -0,0 +1,228 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===---------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines kernels for elementwise evaluation of ROUND(x) function.
+//===---------------------------------------------------------------------===//
+
+#pragma once
+#include <CL/sycl.hpp>
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "kernels/elementwise_functions/common.hpp"
+
+#include "utils/offset_utils.hpp"
+#include "utils/type_dispatch.hpp"
+#include "utils/type_utils.hpp"
+#include <pybind11/pybind11.h>
+
+namespace dpctl
+{
+namespace tensor
+{
+namespace kernels
+{
+namespace round
+{
+
+namespace py = pybind11;
+namespace td_ns = dpctl::tensor::type_dispatch;
+
+using dpctl::tensor::type_utils::is_complex;
+
+template <typename argT, typename resT> struct RoundFunctor
+{
+
+    // is function constant for given argT
+    using is_constant = typename std::false_type;
+    // constant value, if constant
+    // constexpr resT constant_value = resT{};
+    // is function defined for sycl::vec
+    using supports_vec = typename std::false_type;
+    // do both argTy and resTy support sugroup store/load operation
+    using supports_sg_loadstore = typename std::negation<
+        std::disjunction<is_complex<resT>, is_complex<argT>>>;
+
+    resT operator()(const argT &in)
+    {
+        if constexpr (std::is_integral_v<argT>) {
+            return in;
+        }
+        else if constexpr (is_complex<argT>::value) {
+            using realT = typename argT::value_type;
+
+            const realT x = std::real(in);
+            const realT y = std::imag(in);
+            realT x_round, y_round;
+            if (std::abs(x - std::floor(x)) == std::abs(x - std::ceil(x))) {
+                x_round = static_cast<int>(std::ceil(x)) % 2 == 0
+                              ? std::ceil(x)
+                              : std::floor(x);
+            }
+            else {
+                x_round = std::round(x);
+            }
+            if (std::abs(y - std::floor(y)) == std::abs(y - std::ceil(y))) {
+                y_round = static_cast<int>(std::ceil(y)) % 2 == 0
+                              ? std::ceil(y)
+                              : std::floor(y);
+            }
+            else {
+                y_round = std::round(y);
+            }
+            return resT{x_round, y_round};
+        }
+        else {
+            if (in == 0) {
+                return in;
+            }
+            else if (std::abs(in - std::floor(in)) ==
+                     std::abs(in - std::ceil(in))) {
+                return static_cast<int>(std::ceil(in)) % 2 == 0
+                           ? std::ceil(in)
+                           : std::floor(in);
+            }
+            else {
+                return std::round(in);
+            }
+        }
```

Comment:
Factor out real-valued floating-point logic into a private templated method of the struct, and use from here, as well as in the complex branch.

If you ever need to change the logic for the FP type, it would be in a single place.